File: istaroth/agd/renderables/anecdote.py
"""Anecdote (Odd Encounter) processing and rendering."""

from istaroth import utils
from istaroth.agd import (
    first_seen,
    id_types,
    issues,
    localization,
    processed_types,
    repo,
)
from istaroth.agd.renderables import _talk as _talk
from istaroth.text import types as text_types


def get_anecdote_info(
    anecdote_id: id_types.AnecdoteId, *, data_repo: repo.DataRepo
) -> processed_types.AnecdoteInfo | None:
    """Assemble an anecdote's blurbs and storyboard talks, or None if talk-less."""
    entry = data_repo.load_anecdote_excel_config_data()[anecdote_id]
    text_map = data_repo.build_text_map_tracker()
    quest_to_talks = data_repo.build_storyboard_quest_to_talk_ids_mapping()

    talks = list[processed_types.TalkInfo]()
    talk_ids = list[id_types.TalkId]()
    for quest_id in entry["questIds"]:
        for talk_id in quest_to_talks.get(quest_id, []):
            try:
                talk_info = _talk.get_talk_info_by_id(talk_id, data_repo=data_repo)
            except ValueError:
                issues.record(issues.IssueType.MISSING_TALK, str(talk_id))
                continue
            # Require a non-skip line: skip-flagged (dev/test) lines are dropped
            # at render time, so an all-skip talk would emit an empty section.
            if any(not text.skip for text in talk_info.text):
                talks.append(talk_info)
                talk_ids.append(talk_id)

    if not talks:
        return None

    return processed_types.AnecdoteInfo(
        anecdote_id=anecdote_id,
        title=text_map.get_optional(entry["titleTextMapHash"])
        or f"Anecdote {anecdote_id}",
        teaser=text_map.get_optional(entry["teaserTextMapHash"]),
        description=text_map.get_optional(entry["descTextMapHash"]),
        talks=talks,
        talk_ids=talk_ids,
    )


def render_anecdote(
    anecdote: processed_types.AnecdoteInfo,
    *,
    language: localization.Language,
    first_seen_index: first_seen.FirstSeenIndex,
) -> processed_types.RenderedItem:
    """Render an anecdote's blurbs and dialogue into a single file."""
    filename = (
        f"{anecdote.anecdote_id}_{utils.make_safe_filename_part(anecdote.title)}.txt"
    )

    content_lines = [f"# {anecdote.title}\n"]
    for blurb in (anecdote.teaser, anecdote.description):
        if blurb is not None:
            content_lines.append(blurb)
            content_lines.append("")

    for i, talk in enumerate(anecdote.talks):
        content_lines.append(f"## Talk {i}\n")
        content_lines.extend(_talk.render_talk_content(talk, language))
        content_lines.append("")

    min_version, max_version = first_seen_index.resolve(
        [
            first_seen.SourceId(first_seen.SourceDomain.TALK, talk_id)
            for talk_id in anecdote.talk_ids
        ]
    )
    return processed_types.RenderedItem(
        text_metadata=text_types.TextMetadata(
            category=text_types.TextCategory.AGD_ANECDOTE,
            title=anecdote.title,
            id=anecdote.anecdote_id,
            relative_path=f"{text_types.TextCategory.AGD_ANECDOTE.value}/{filename}",
            min_version=min_version,
            max_version=max_version,
        ),
        content="\n".join(content_lines).rstrip(),
    )

File: rust/istaroth-agd-regen/src/renderables/anecdote.rs
//! Port of renderables/anecdote.py: Anecdote (Odd Encounter) blurbs and
//! storyboard talks.

use crate::firstseen::Domain;
use crate::issues::{IssueType, Scope};
use crate::renderables::talk::{self, TalkInfo, TalkNotFound};
use crate::rendered_item::RenderedItem;
use crate::repo::Repo;
use crate::util;
use crate::vh::{ValueExt, int_array};
use anyhow::{Result, anyhow};

/// Anecdotes pass discovery: sorted anecdote ids.
pub fn discover(repo: &Repo) -> Result<Vec<i64>> {
    let mut ids: Vec<i64> = repo.excel.anecdote.keys().copied().collect();
    ids.sort();
    Ok(ids)
}

/// Assemble an anecdote's blurbs and storyboard talks, or None if talk-less.
/// `has_non_skip_text` is required: skip-flagged (dev/test) lines are dropped
/// at render time, so an all-skip talk would emit an empty section.
pub fn process(repo: &Repo, scope: &Scope, anecdote_id: i64) -> Result<Option<RenderedItem>> {
    let entry = repo
        .excel
        .anecdote
        .get(&anecdote_id)
        .ok_or_else(|| anyhow!("unknown anecdote {anecdote_id}"))?;
    let mut talks: Vec<TalkInfo> = Vec::new();
    let mut talk_ids: Vec<i64> = Vec::new();
    for quest_id in int_array(entry.f("questIds")?)? {
        for talk_id in repo
            .storyboard_quest_to_talk_ids
            .get(&quest_id)
            .into_iter()
            .flatten()
        {
            let talk_info = match talk::get_talk_info_by_id(repo, scope, *talk_id) {
                Ok(info) => info,
                Err(e) if e.is::<TalkNotFound>() => {
                    scope.record_issue(IssueType::MissingTalk, talk_id.to_string());
                    continue;
                }
                Err(e) => return Err(e),
            };
            if talk_info.has_non_skip_text() {
                talks.push(talk_info);
                talk_ids.push(*talk_id);
            }
        }
    }
    if talks.is_empty() {
        return Ok(None);
    }

    let title = repo
        .tm
        .get_optional(entry.i("titleTextMapHash")?, scope)?
        .filter(|t| !t.is_empty())
        .unwrap_or_else(|| format!("Anecdote {anecdote_id}"));
    let teaser = repo.tm.get_optional(entry.i("teaserTextMapHash")?, scope)?;
    let description = repo.tm.get_optional(entry.i("descTextMapHash")?, scope)?;

    let filename = format!(
        "{anecdote_id}_{}.txt",
        util::make_safe_filename_part(&title)
    );
    let mut content_lines = vec![format!("# {title}\n")];
    for blurb in [&teaser, &description].into_iter().flatten() {
        content_lines.push(blurb.clone());
        content_lines.push(String::new());
    }
    for (i, talk_info) in talks.iter().enumerate() {
        content_lines.push(format!("## Talk {i}\n"));
        content_lines.extend(talk::render_talk_content(talk_info, repo.language, scope)?);
        content_lines.push(String::new());
    }
    let versions = repo.first_seen.resolve_ints(Domain::Talk, talk_ids)?;
    Ok(Some(RenderedItem::new(
        "agd_anecdote",
        title,
        anecdote_id,
        filename,
        versions,
        util::py_rstrip(&content_lines.join("\n")).to_string(),
    )))
}
